### PR TITLE
Add SHARD_GRAD_OP to device mesh error check

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Add SHARD_GRAD_OP to device mesh error check. Avoids nasty torch error if user incorrectly specifies mesh shard combo

### CODE REVIEW

File: composer/trainer/dist_strategy.py [hunk 1]
```diff
@@ -303,7 +303,9 @@ def sync_hook(*args):
     if version.parse(torch.__version__.split('.dev')[0]) >= version.parse('2.2.0'):
         if 'device_mesh' in fsdp_config:
             device_mesh_size = len(fsdp_config['device_mesh'])
-            if sharding_strategy in [ShardingStrategy.FULL_SHARD, ShardingStrategy.NO_SHARD] and device_mesh_size != 1:
+            if sharding_strategy in [
```
Comment (b-chu) [inline composer/trainer/dist_strategy.py:306]: Add parentheses for better formatting

File: tests/trainer/test_fsdp.py [hunk 2]
```diff
@@ -219,10 +220,19 @@ def test_fsdp_process_group(world_size: int):
 @pytest.mark.gpu
 @world_size(2)
 @pytest.mark.skipif(version.parse(torch.__version__) < version.parse('2.2.0'), reason='Device mesh requires Torch 2.2')
-def test_wrong_size_device_mesh_error(world_size: int):
-    with pytest.raises(ValueError, match='.*requires a device mesh of size 1.*'):
+@pytest.mark.parametrize('sharding_strategy',
+                         ['NO_SHARD', 'SHARD_GRAD_OP', 'FULL_SHARD', 'HYBRID_SHARD', '_HYBRID_SHARD_ZERO2'])
```
Comment (b-chu) [inline tests/trainer/test_fsdp.py:224]: Add trailing comma

**Review by b-chu — APPROVED** [review]

File: composer/trainer/dist_strategy.py (hunk 1 above)
Comment (mvpatel2000) [reply composer/trainer/dist_strategy.py:306]: Discussed offline, looks ugly regardless

File: tests/trainer/test_fsdp.py (hunk 2 above)
Comment (mvpatel2000) [reply tests/trainer/test_fsdp.py:224]: Discussed offline, looks ugly regardless

**Pull request merged**